pkg/util/kubernetes/apiserver/leaderelection/leaderelection.go

```diff
@@ -228,9 +228,15 @@ func (le *LeaderEngine) EnsureLeaderElectionRuns() error {
 
 func (le *LeaderEngine) runLeaderElection() {
 	for {
-		log.Infof("Starting leader election process for %q...", le.HolderIdentity)
-		le.leaderElector.Run(le.ctx)
-		log.Info("Leader election lost")
+		select {
+		case <-le.ctx.Done():
+			log.Infof("Quitting leader election process: context was cancelled")
+			return
+		default:
+			log.Infof("Starting leader election process for %q...", le.HolderIdentity)
+			le.leaderElector.Run(le.ctx)
```

Comment:
The `Run` method of the leader elector is blocking, and will terminate if `ctx` was cancelled OR if the current instance lost leadership.

In case we lost leadership, we should call `Run` again to participate in the leader election process. 

However, if the context was cancelled, we should just quit.

Reply:
From the documentation: 

```
Run will not return before leader election loop is stopped by ctx or it has stopped holding the leader lease
```